src/timer.h

```diff
@@ -54,69 +59,79 @@ typedef struct timespec hr_timer;
-#define HRZERO(X)             (X).tv_sec = (X).tv_nsec = 0
-#define HRSETCURRENT(X)       clock_gettime(BEST_MONOTONIC, &(X))
-#define HRGETTICKS(X)         ((uint64_t)(X).tv_sec * 1000000000ULL + (uint64_t)(X).tv_nsec)
-#define HRGETTICKS_PER_SEC(X) { hr_timer r; clock_getres(BEST_MONOTONIC, &r); if (r.tv_sec < 0) clock_getres(CLOCK_MONOTONIC, &r); (X) = r.tv_sec + 1000000000ULL / r.tv_nsec; }
 
-#else /* Fallback to microsecond non-monotonic clock that should be available */
+#define HRZERO(X)         (X).tv_sec = (X).tv_nsec = 0
+#define HRSETCURRENT(X)   clock_gettime(BEST_MONOTONIC, &(X))
+#define HRGETTICKS(X)     ((uint64_t)(X).tv_sec * 1000000000ULL + (uint64_t)(X).tv_nsec)
+#define HRSETRESOLUTION() (timer_ticks_per_sec = 1000000000ULL)
+#define HRTICKS_PER_SEC   (1000000000ULL)
+#define timer_get_nano()  ({ hr_timer t; HRSETCURRENT(t); HRGETTICKS(t); })
+
+#else /* Fallback to microsecond non-monotonic clock that should always be available */
 
 #include <sys/time.h>
 
+/*
+ * gettimeofday gives s + us
+ */
+#define HR_CLOCK          "gettimeofday()"
 typedef struct timeval hr_timer;
-#define HRZERO(X)             (X).tv_sec = (X).tv_usec = 0
-#define HRSETCURRENT(X)       gettimeofday(&(X), NULL)
-#define HRGETTICKS(X)         ((uint64_t)(X).tv_sec * 1000000 +	\
-                               (uint64_t)(X).tv_usec)
-#define HRGETTICKS_PER_SEC(X) (X) = 1000000
+
+#define HRZERO(X)         (X).tv_sec = (X).tv_usec = 0
+#define HRSETCURRENT(X)   gettimeofday(&(X), NULL)
+#define HRGETTICKS(X)     ((uint64_t)(X).tv_sec * 1000000 + (uint64_t)(X).tv_usec)
+#define HRSETRESOLUTION() (timer_ticks_per_sec = 1000000)
+#define HRTICKS_PER_SEC   (1000000)
+#define timer_get_nano()  ({ hr_timer t; HRSETCURRENT(t); HRGETTICKS(t) * 1000; })
 
 #endif
 
-#endif /* Windows or Unix */
-
-typedef struct sTimer_s {
-	int m_fRunning;     // true if we are running
-	clock_t m_cStartTime;
-	clock_t m_cEndTime;
-	hr_timer m_hrStartTime;
-	hr_timer m_hrEndTime;
-	double m_dAccumSeconds;
-} sTimer;
-
-extern void sTimer_Init(sTimer *t);      // Init
-extern void sTimer_Start(sTimer *t, int bClear /*=true*/);  // Start the timer
-extern void sTimer_Stop(sTimer *t);      // Stop the timer
-extern void sTimer_ClearTime(sTimer *t); // Clears out the time to 0
-extern double sTimer_GetSecs(sTimer *t); // If timer running returns elapsed;
-                                         // if stopped returns timed interval;
-                                         // if not started returns 0.
-
-extern uint64_t sm_HRTicksPerSec;  // HR ticks per second (claimed)
-extern uint64_t sm_hrPrecision;    // HR ticks per second (observed, guess)
-extern uint64_t sm_cPrecision;     // clocks (ticks) per second (observed, guess)
-
-#define sTimer_Start_noclear(t) \
+#endif /* defined (_MSC_VER) || defined (__MINGW32__) || defined (__CYGWIN32__) */
+
+typedef struct hr_timer_s {
+	int running;
+	hr_timer start;
+	hr_timer end;
+	uint64_t accumulated;
+} hr_timer_t;
```

Comment:
I don't suggest we necessarily change it, but JFYI per POSIX names `*_t` are reserved for use by the system. This is one of the most frequently violated rules, and we should too, but it could be wise to use names that are unrealistic to clash with anything a system might introduce, e.g. `john_timer_t`.